### ISSUE TITLE
Unexpected performance of coalton while running `(microbench1:run)`

### ISSUE THREAD

**Issue by digikar99** [coalton-lang/coalton#985]
Hello,

Something seems off regarding performance and inlining in the current master branch version of coalton, or may be I'm missing something. 

I started SBCL with `rlwrap sbcl --no-userinit --load quicklisp/setup.lisp`:

```lisp
* (asdf:load-system "coalton" :force t)
...
* (ql:quickload "small-coalton-programs")
To load "small-coalton-programs":
  Load 1 ASDF system:
    small-coalton-programs
; Loading "small-coalton-programs"
[package small-coalton-programs]..................
[package small-coalton-programs.primes-native]....
[package small-coalton-programs.primes-iterator]..
[package freecalc]................................
[package lisp-microbench1]........................
[package coalton-microbench1].....................
[package microbench1].
("small-coalton-programs")
* (microbench1:run)
Lisp took 37084 ms
Coalton took 25723 ms
Coalton was faster by 11361 ms (30.64%)
The hand-optimized function took 8020 ms
NIL
```

I expected coalton to take about the same time as the hand-optimized function, but it took significantly longer and seems closer to the lisp version than the hand-optimized version. (I also set the minimum and maximum clock speeds for my CPU at 0.8GHz.)

So, I checked the disassembly of the `add-two-caller` below, and it turns out to be not inlined:

```lisp
(in-package :coalton-user)

(named-readtables:in-readtable coalton:coalton)

(coalton-toplevel
  (declare add-two (single-float -> single-float))
  (define (add-two x) (+ x 2.0)))

(cl:defun add-two-caller (x)
  (cl:declare (cl:optimize cl:speed (cl:safety 0) (cl:debug 0))
              (cl:type cl:single-float x))
  (add-two x))
```

```lisp
(cl:disassemble 'add-two-caller)
; disassembly for ADD-TWO-CALLER
; Size: 13 bytes. Origin: #x54078956                          ; ADD-TWO-CALLER
; 56:       B902000000       MOV ECX, 2
; 5B:       FF7508           PUSH QWORD PTR [RBP+8]
; 5E:       E9DFCE36FC       JMP #x503E5842                   ; #<FDEFN ADD-TWO>
```

Am I missing something here - or is this a (known) bug?

**Comment by stylewarning** [comment]
Did you try building with [`:coalton-release`](https://github.com/coalton-lang/coalton/blob/main/docs/coalton-lisp-interop.md#interaction-mode) on your `*features*`?

**Comment by stylewarning** [comment]
The Lisp functions that Coalton makes aren't presently declared as inline.

Inlining in Coalton is a huge area of interest that we haven't tackled much at all, but it will ultimately be critical for high-speed, non-consing numerical code.

(Also, reminder about [`:coalton-release`](https://github.com/coalton-lang/coalton/blob/main/docs/coalton-lisp-interop.md#interaction-mode).)

**Comment by stylewarning** [comment]
Oh, and second, the performance of hand-optimized Lisp being better than Coalton is a known area of improvement. (I hesitate to call it a bug, because it's not a regression, and it still calculates the correct thing.)

**Comment by digikar99** [comment]
> Did you try building with [:coalton-release](https://github.com/coalton-lang/coalton/blob/main/docs/coalton-lisp-interop.md#interaction-mode) on your `*features*`?

No, I hadn't - thanks for the pointer! Although, even with this, I don't see much performance improvement.

So, that leads me to two thoughts - 

## 1. About the separation between development and release mode

Is there something that makes this separation necessary? Or is the merger something that is still to be thought and worked out?

> ### Development Mode
> 
> Development mode allows most structures to be re-defined. This means:
> 
> * Most types turn into CLOS classes.
> * Most types are not flattened or unwrapped in any way.
> * Certain optimizations that obscure debugging are disabled.
> 
> ### Release Mode
> 
> Release mode freezes most structures and optimizes them.
> 
> * Most types are defined as frozen `defstruct`s or flattened Lisp data types.
> * Many optimizations are performed.

If I understand the concern about type safety, I'm guessing that the frozen types during release mode aid type safety by  ensuring that the types don't get redefined after their first definition. But on the other hand, this hinders the rapid prototyping capabilities of lisp which involve redefinitions upon redefinitions, which make the development mode a necessity too. Is there something else that makes the separation necessary?

But that said, type safety seems a bit independent to whether or not a function is available for inspection during run-time. Regarding the latter, it should be possible to `(optimize debug)` in both cases, so that the function calls (in coalton or in lisp) do not disappear from the call stack by inlining; or `(optimize speed)`, so that the function calls may or may not remain on the call stack.

This should also be relevant to https://github.com/coalton-lang/coalton/discussions/978.

## 2. Inlining

> The Lisp functions that Coalton makes aren't presently declared as inline.
> 
> Inlining in Coalton is a huge area of interest that we haven't tackled much at all, but it will ultimately be critical for high-speed, non-consing numerical code.

Yes, this would be a deal breaker for anyone who wants to provide coalton-based libraries to both coalton and non-coalton users :(. At the outset, I don't see anything inherent that might prevent inlining coalton code into lisp, so I might look into it in the upcoming weeks. I do suspect this is going to require compiler-macros - and perhaps [cl-form-types](https://github.com/alex-gutev/cl-form-types/) too to optimize and inline the non-trivial cases, and thus the CLTL2 environment API. But okay, that's going too fast, and as a first step we do need to get inlining working in the trivial cases.

> Oh, and second, the performance of hand-optimized Lisp being better than Coalton is a known area of improvement.

Hmm, interesting. Though I do expect it would be easier to optimize coalton than lisp. And perhaps the difference should vanish once the inlining gets fixed.

**Comment by eliaslfox** [comment]
> Is there something that makes this separation necessary? Or is the merger something that is still to be thought and worked out?

The development/release mode divide isn't about type safety, it's about allowing both interactive development and performance. When you define a CLOS class, the accessors compile to something like:

```common-lisp
(defclass point () 
  ((x :accessor point-x)
   (y :accessor point-y)))

(defmethod ((obj pointj))
  (slot-value point 'x))
```

Each field access pays the cost of a CLOS method lookup, and then the cost of slot-value. Structs on the other hand have a fixed layout which can make field access much faster. 

```common-lisp
CL-USER> (declare (optimize (speed 3) (safety 0)))
CL-USER> (defstruct point x y)

CL-USER> (disassemble 'point-x)
; disassembly for POINT-X
; Size: 20 bytes. Origin: #x700CFD6BF8                        ; POINT-X
; BF8:       4AD140F8         LDR R0, [R0, #13]

; BFC:       FB031AAA         MOV CSP, CFP
; C00:       5A7B40A9         LDP CFP, LR, [CFP]
; C04:       BF0300F1         CMP NULL, #0
; C08:       C0035FD6         RET
```
OTOH the fixed field layout of structs makes safely redefining them generally impossible. Compiled code that accuses fields needs to be recompiled if the location of a field changes.

> I do suspect this is going to require compiler-macros - and perhaps [cl-form-types](https://github.com/alex-gutev/cl-form-types/) too to optimize and inline the non-trivial cases

Toplevel Coalton functions compile to defuns. You can just inline them with `(declare (inline function-name))`.

**Comment by eliaslfox** [comment]
In this specific microbenchmark the problem appears that + and * for floats aren't being inlined in the Coalton version. We used to inline every directly called method but I removed it because it broke redefinition. Some of the base methods in the standard library should probably be explicitly marked inline.

**Comment by digikar99** [comment]
> The development/release mode divide isn't about type safety, it's about allowing both interactive development and performance.

Oh, I see. Or, erm, no, I don't.

I had mistook the structures and classes to be representing the compile time type information required for the compilation of coalton code. But on a careful reading, it seems that they also exist at runtime just like most things in the lisp world. So, your clarification helps!

The part I don't get then is this - and this might actually go on a tangent and so might deserve a separate issue/discussion: if coalton types are the same as lisp structures/classes at run time, and lisp structures (at least on SBCL) are non-zero cost abstractions, then does that mean coalton types are also non-zero cost abstractions? If so, have zero-cost abstractions been in discussion in the context of coalton?

Regarding inlining,

> Toplevel Coalton functions compile to defuns. You can just inline them with `(declare (inline function-name))`.

While this literally inlines them, it doesn't seem to actually inline them in a useful way. So, for example, in the below code - 

```lisp
(in-package :coalton-user)

(named-readtables:in-readtable coalton:coalton)

(cl:declaim (cl:inline add-two))

(coalton-toplevel  
  (monomorphize)
  (declare add-two (single-float -> single-float))
  (define (add-two x) (+ x 2.0)))

(cl:defun add-two-caller (x)
  (cl:declare (cl:optimize cl:speed (cl:safety 0) (cl:debug 0))
              (cl:type cl:single-float x))
  (add-two x))
```

The disassembly of `add-two-caller`  (or even the `add-two`) actually includes a call to another function, and so the addition of two single floats is still a non-inline operation.

```
COALTON-USER> (disassemble 'add-two-caller)
; disassembly for ADD-TWO-CALLER
; Size: 31 bytes. Origin: #x5401A586                          ; ADD-TWO-CALLER
; 86:       4883EC10         SUB RSP, 16
; 8A:       488B3DC7FFFFFF   MOV RDI, [RIP-57]                ; 2.0
; 91:       B904000000       MOV ECX, 4
; 96:       48892C24         MOV [RSP], RBP
; 9A:       488BEC           MOV RBP, RSP
; 9D:       E820763CFC       CALL #x503E1BC2                  ; #<FDEFN COALTON-LIBRARY/MATH/NUM::|INSTANCE/NUM SINGLE-FLOAT-COALTON-LIBRARY/CLASSES:+|>
; A2:       C9               LEAVE
; A3:       F8               CLC
; A4:       C3               RET
COMMON-LISP:NIL
```

**Comment by eliaslfox** [comment]
> The part I don't get then is this - and this might actually go on a tangent and so might deserve a separate issue/discussion: if coalton types are the same as lisp structures/classes at run time, and lisp structures (at least on SBCL) are non-zero cost abstractions, then does that mean coalton types are also non-zero cost abstractions? 

Types don't need to have a runtime cost, but constructing values almost always will. All of the following have equivalent runtime cost:

```
(define-type (T :a) (T String))

(define x (the (T Unit) (T "x)))

(define y (the (T (List Integer)) (T "y")))

(define z (the (T (List (List Integer))) (T "z")))
```

> If so, have zero-cost data abstractions been in discussion in the context of coalton?

How would this work?

> While this literally inlines them, it doesn't seem to actually inline them in a useful way. 

Inlining call sites recursively is a difficult problem. It's easy to write a compiler pass to do it, but figuring out heuristics for which functions should be inlined is hard. Recursive inlining would also break the redefinition in the REPL. Coalton's inlining is on par with common lisp's (inline when told to) but much worse than the state of the art (global heuristic based inlining w/ LTO.) I don't think it's an unsolvable problem, but there's a lot of details to work out.

**Comment by stylewarning** [comment]
> Toplevel Coalton functions compile to defuns. You can just inline them with (declare (inline function-name)).

@eliaslfox A DEFUN can only be inlined if it was declared INLINE before the definition. The pattern to make a function inlineable is

```
(declaim (inline f))
(defun f ...)
(declaim (notinline f))
```

This tells CL to save inlining info for F so as to be explicitly inlined later.

**Comment by digikar99** [comment]
> Inlining call sites recursively is a difficult problem. It's easy to write a compiler pass to do it, but figuring out heuristics for which functions should be inlined is hard. Recursive inlining would also break the redefinition in the REPL. Coalton's inlining is on par with common lisp's (inline when told to) but much worse than the state of the art (global heuristic based inlining w/ LTO.) I don't think it's an unsolvable problem, but there's a lot of details to work out.

I'm not referring to the heuristics themselves. May be there's a better word than "inlining" to refer to what I have in mind: I am okay with adding certain declarations or anything, but the end result I'm looking for is that the `disassemble` of `add-two-caller` should contain the assembly instruction `ADDSS` when the code is compiled, using say SBCL, on an intel CPU. At the moment, the disassembly instead contains a call to `COALTON-LIBRARY/MATH/NUM::|INSTANCE/NUM SINGLE-FLOAT-COALTON-LIBRARY/CLASSES:+` instead of a simple `ADDSS` instruction.

---

> > If so, have zero-cost data abstractions been in discussion in the context of coalton?
> How would this work?

I'm not sure how although I do have some ideas, but just to clarify - I should have written zero run-time cost data abstraction. For instance, when the following C code is compiled using `gcc`, then the disassembly of `do_tf` and `do_raw` are identical, no runtime cost is encoutered by using a `twin_float` structure/type:

```c
struct twin_float{double a; double b;};

double do_tf(int n){
  double sum = 0;
  for(int i=0; i<n; i++){
    struct twin_float tf = {(double)i, (double)i};
    sum += tf.a;
    sum += tf.b;
  }
  return sum;
}

double do_raw(int n){
  double sum = 0;
  for(int i=0; i<n; i++){
    double a = (double)i;
    double b = (double)i;
    sum += a;
    sum += b;
  }
  return sum;
}
```

However, when the following code is compiled using SBCL 2.3.8 on a 64-bit intel processor, the disassembly of `do-tf` and `do-raw` are actually different - the code for `do-tf` actually contains instructions to create a `twin-float` value. And in this particular case, the runtime performance impact is non-existent to trivial at best, I still wonder if there are existing lisp/coalton efforts to address this.

```lisp
(declaim (inline make-twin-float))
(defstruct (twin-float (:constructor make-twin-float (a b)))
  (a 0.0f0 :type double-float)
  (b 0.0f0 :type double-float))

(defun do-tf (n)
  (declare (optimize speed)
           (type double-float n))
  (let ((sum 0.0d0))
    (declare (type double-float sum))
    (loop :for i :from 0.0d0 :below n :by 1.0d0
          :do  (let ((tf (make-twin-float i i)))
                 (declare (dynamic-extent tf))
                 (incf sum (twin-float-a tf))
                 (incf sum (twin-float-b tf))))
    sum))

(defun do-raw (n)
  (declare (optimize speed)
           (type double-float n))
  (let ((sum 0.0d0))
    (declare (type double-float sum))
    (loop :for i :from 0.0d0 :below n :by 1.0d0
          :do (let ((a i)
                    (b i))
                (incf sum a)
                (incf sum b)))
    sum))
```

As I understand, it is the following assembly code in `do-tf` that indicates the allocation of `twin-float` :

```
; 4E:       C70039080000     MOV DWORD PTR [RAX], 2105
; 54:       C74004039B2D50   MOV DWORD PTR [RAX+4], 1345166083
```

**Comment by digikar99** [comment]
As I understand, the issue above is that `+` itself gets inlined, but its instances do not get inlined.

I am looking at the coalton code, and I haven't yet been able to understand how a toplevel `define-instance` form gets macroexpanded. A proposal was this: If a `function-name` has been declaimed inline previously, then the corresponding function instances created by `define-instance` also get declaimed inline. Later on, the `function-name` may be declaimed `notinline` to prevent inlining by default. 

This way, whenever a user writes `(declare (inline function-name))`, then, not only is `function-name` inlined, but its appropriate instance is inlined too. In the case of `+` above, the instance is `COALTON-LIBRARY/MATH/NUM::|INSTANCE/NUM SINGLE-FLOAT-COALTON-LIBRARY/CLASSES:+|`.

**Comment by swapneils** [comment]
Since `coalton` is our own interface layer and (I believe) we already walk the code when converting it to CL, can we track toplevel functions as objects and make a calling graph of function invocations, so that we can inline all `coalton` functions and just recompile already-compiled callers when the callee is redefined? 

It's more orchestration overhead, but that should be regarding compilation time, not runtime.

**Comment by eliaslfox** [comment]
Yeah you could do that. If you changed the type signature of a function then you would need to re-typecheck all call sites.

**Comment by dieggsy** [comment]
I'm not sure if this is related (and sorry if this is a silly question, it's been a good while since I dealt with a functional/ML-style language, and I never really went that deep)

I was going through [Project Euler problem 2](https://projecteuler.net/problem=2) with Coalton and ran into a performance discrepancy I didn't expect. This is with `(coalton-impl/settings:coalton-release-p)` returning `t`

Here's my lisp code and a `time`:
```lisp
(defun scheme-translation ()
  (labels ((lp (i j res)
             (if (>  j 4000000)
                 res
                 (lp j (+ i j) (if (evenp j) (+ res j) res)))))
    (lp 1 2 0)))

(time (loop :repeat 10000000 :do (scheme-translation)) )
```
```
Evaluation took:
  1.160 seconds of real time
  1.160108 seconds of total run time (1.160084 user, 0.000024 system)
  100.00% CPU
  4,515,836,988 processor cycles
  0 bytes consed
```

Here's my Coalton code:
```lisp
(coalton-toplevel
  (define (scheme-translation)
    (rec lup ((i 1)
              (j 2)
              (sum 0))
      (cond ((> j 4000000)
             sum)
            (true
             (lup j (+ i j)
                  (if (even? j)
                      (+ sum j)
                      sum)))))))

(time (loop :repeat 10000000 :do (coalton:coalton (scheme-translation))) )
```
```
Evaluation took:
  8.371 seconds of real time
  8.369100 seconds of total run time (8.368883 user, 0.000217 system)
  99.98% CPU
  32,584,749,756 processor cycles
  0 bytes consed
```

The fully type-inferred Coalton with no explicit type declarations was significantly slower than the Lisp with no type declarations. When I declare J to be an integer (I just put a `(the integer j)` at the top of the rec body), it catches up (is consistently a little faster, even):

```
Evaluation took:
  0.814 seconds of real time
  0.814038 seconds of total run time (0.807035 user, 0.007003 system)
  100.00% CPU
  3,168,751,521 processor cycles
  13,872 bytes consed
```

I sort of gleaned some information from the Coalton type of `scheme-translation` before and after:
```
∀ :A :B. COALTON-LIBRARY/MATH/INTEGRAL:INTEGRAL :B ⇒ (:A → :B)
```
∀ :A. (:A → COALTON:INTEGER)
```

Is the discrepancy in the first case just because Coalton also has higher level, not-as-performant types than Lisp (the generic `Integral` here?), and without any declarations it can't specialize any further so operations are very generic, but Lisp can't do any worse than adding reals in this case?

**Comment by stylewarning** [comment]
@dieggsy I haven't tried your example so I won't be able to give you a precise answer, but you're essentially right. Coalton's Integral types don't even need to be subtypes of `cl:integer`; they can be whatever type is declared to have Integral-like semantics. As such, the fully general code is much more general and extensible than the Common Lisp code. Declaring a more specific type allows Coalton to generate much better code. Even more so if you choose UFix over Integer.

**Comment by stylewarning** [comment]
Current benchmark:

```
CL-USER> (microbench1:run)
Coalton was compiled in DEVELOPMENT mode.
Lisp took 5511 ms
Coalton took 1563 ms
    => Coalton was faster by 3948 ms (71.64%)
The hand-optimized function took 1644 ms
    => Coalton was faster by 81 ms (4.93%)
```

I think we can close this issue now. :)
